dialogue-core/src/main/java/com/palantir/dialogue/core/QueuedChannel.java

```diff
@@ -179,7 +199,24 @@ private boolean scheduleNextTask() {
                         DIRECT);
                 return true;
             } else {
-                queuedCalls.addFirst(components);
+                if (!queuedCalls.offerFirst(components)) {
+                    // Should never happen, ConcurrentLinkedDeque has no maximum size
+                    log.error(
+                            "Failed to add an attempted call back to the deque",
+                            SafeArg.of("service", endpoint.serviceName()),
+                            SafeArg.of("endpoint", endpoint.endpointName()));
+                    queueSizeEstimate.decrementAndGet();
+                    if (!queuedResponse.setException(
+                            new SafeRuntimeException(
+                                    "Failed to req-queue request",
+                                    SafeArg.of("service", endpoint.serviceName()),
+                                    SafeArg.of("endpoint", endpoint.endpointName())))) {
+                        log.debug(
+                                "Queued response has already been completed",
+                                SafeArg.of("service", endpoint.serviceName()),
+                                SafeArg.of("endpoint", endpoint.endpointName()));
+                    }
+                }
                 return false;
```

Comment:
you know the thing we talked about in today's zoom, where if we can't make progress on the head request due to endpoint-level blacklisting, then the entire queue can't make progress...

if we just iterated through the whole queue here rather than just stopping on the first one what was limited, wouldn't this problem go away? (Could use the weird iterator#remove method if it succeeds)

Reply:
We attempt to schedule a _lot_. In normal operation, the queue is empty so there's nothing to worry about. When channels are limiting most requests, turning this from an O(1) operation to an O(n) operation would be problematic.